### ISSUE TITLE
 Spike: Pull latest published resource from Catalog Next

### ISSUE THREAD

**Issue by birchamp** [unfoldingWord/tc-create-app#829]
All new spec'd TSVs will use the Catalog Next release tags instead of being published to Door43-catalog.

Can we use GRT to handle Catalog Next features? Particularly release tags.
How will it handle repo objects?

**Comment by richmahn** [comment]
If GRT returns a repo object as is from DCS/Gitea's API, then you should have information about ever "stage" of the catalog, if you want to show the latest release (prod), "pre-prod" (set in stone but not yet prod, a setting when making a release), "draft" (not yet set in stone but points to the end of a given branch), and "latest" (master branch if manifest.yaml is valid). 

See https://git.door43.org/api/v1/repos/unfoldingword/en_obs as an example. The "catalog" property has an object and entry for each of these stages to show if the repo has them and what the name of that release is (branch_or_tag_name) and then a API URL for the release.

**Comment by richmahn** [comment]
Ok, GRT DOES return the catalog/release info of a repo, as one can verify here: https://gitea-react-toolkit.netlify.app/#/Repositories%20?id=repositories-1

The search results show a JSON (all on one line) so you have to scroll to the right and see the catalog entries.

**Comment by mandolyte** [comment]
## Links

The Catalog Next overview:
https://forum.door43.org/t/catalog-next-populating-accessing-the-dcs-resource-catalog/899/8

The Catalog Next Swagger:
https://qa.door43.org/api/catalog/swagger#/v5/v5SearchOwner

## Prior Work

1. The caching proposal below is similar to the approach used by the Content Validation App, which was developed by Bruce McLean.
2. The `gatewayEdit` codebase includes a hook that wraps the NPM "Local Forage" component and manages data in the browser's `indexedDB` database.

## Proposal - High Level

Create a headless component library that integrates CN with DCS.

The features of this CN component are:

1. For production users, restrict application access to production releases for our resources.
2. For developers, allow specification of the CN release to use for resources
3. For improved performance (and reduced network and server resources), provide automatic management of caching based on resource releases. A description of the caching features is below.

## Proposal - Details

### Initialization

The CN component requires initialization when the application starts. This initialization will proceed in the background, so that app is not required to wait until it is done. 

The initialization parameters are:
- SERVER: this points to the copy of DCS to use. Default will be 
`https://qa.door43.org`; for production: `https://git.door43.org`.
- OWNER: this is the owner of the resources. The default is 'unfoldingWord'. 
- LANGID: this provides the language to be used in identifying resources. Default will be `en`.
- RESOURCES: this is a comma delimited list of resource IDs. The complete list is:
'ta', 'tw', 'twl', 'tn', 'tq', 'obs', 'obs-tq', 'obs-tn', 'obs-sn', 'obs-sq'. *Issue: how to specify Original Language resources (Greek and Hebrew texts)*.
- STAGE: this is the value for CN component which determines the release level to work with. It must be one of the following: 'latest', 'preprod', 'prod', 'draft'. The default is 'latest'.
- APIVERSION: this may be hard coded for now to be `api/catalog/v5`.

### API

The CN component will provide a set of functions to access the content of a resource.

Given a server, owner, language, resource type list, and stage:
- initialize(): will run `initalizeCache()` for each resource type asynchronously. Before doing this will it clear the "unzipped cache".

- initializeCache(): will retrieve the corresponding Zip file for the resource.
- If stage is "latest" (default), then...
	- It will query DCS Catalog Next to check what the latest release is.
	- If the resource zip file is already cached. then it does nothing. 
	- If not cached, then it must fetch the "resource zip" from CN and store it in the cache.
- If stage is not "latest", then...
	- It will query DCS Catalog Next to find the requested release
	- If none qualify, it will throw an error. (Note that use of anything other than latest will generally be done by developers.)
	- If a qualifying resource is found, it the process above is followed. (*Issue: what are the "rules" for a qualifying resource?*)

Given an owner, language, and resource type:
- getManifest(): return the manifest as a JSON object. *This would be used to access arbitrary attributes in the manifest.*
- getProjects(): return the list of projects from the manifest. *This would be used to provide a list of projects from which the user may select.*
- getFilenames(): return a list of the files that correspond to the project(s) in the sort order (if any) provided by the manifest. *This would be used to provide a file list from which the user may select.* NOTE: for non-Scripture resources (i.e., TW and TA), a list of filenames will have to be extracted from the zip itself. This list should itself be cached.

Given an owner, language, resource type, and a filename:
- getContent(): returns the content of the file. This function may throw an error if the content does not pass validation.

## Caching Features

There are two levels of caching:
- one that is cleared at the start of the app: "unzipped cache"
- one that contains the released resource CN zip files: "zip cache"

When the app requests a file from the cache (either the manifest or a project file), then:

- The unzipped cache is checked first and the file is returned if found
- If not found, then the file is extracted from the zip file and stored in the unzipped cache; then it is returned.

There may be times when a user logs in and gets to the file selection step before the initialization has fetched the zip file. In this case, the app must show an "in process" spinner until the zip is present for use.

*Issue*: Running apps that use the CN component in mulitple tabs of the same browser may interfere with each other at the cache level. Consider when a new latest is available to the app in the second tab that wasn't available when the app started in the first tab.

**Comment by mandolyte** [comment]
@richmahn could you review the above proposal for Catalog Next accuracy? Thanks.

@PhotoNomad0 I give you credit for the caching scheme in this proposal. Could you review for any flaws in how I described it? Thanks.

**Comment by mandolyte** [comment]
Here is the write up on "editor" vs "translator" workflow.
# Edit vs Translation

The tc-create app has both a "source" and a "target" for content. The source side is on the left side of the page and the target on the right. The term "edit" and "editor" are used to describe those who maintain the English language content of the uW resources. 
- *For editors, both source and target are the exact same repo.* 
- *For translators, the source and target are never the same repo; and generally never the same owner. In other words, translator work is saved in a different org.*

Below are both current process and future catalog next based process.

## Current Process

### Translator Chronology

1. Translator selects a file to translate from the list of files from the source uW master repo branch.
2. Branch is created in target repo and the file is created in that branch populated with the English content from the  uW source master repo branch.
3. The source side is populated from the uW source master branch.
4. The target side may be changed and saved to the translator's branch until done.
5. Translator finishes, saves, and a PR is used to merge this new file into the target master branch.
6. Later, the translator wants to "edit" the translated file.
7. Again the file is selected from the list of files from the  uW source master repo branch.
8. This time, since the target master branch has content for this file, it will be used to populate the translator's branch.
9. Changes are made, saved, and a PR is used to merge the updates back to the target master branch.


### Editor Chronology

This case describes an update to the uW *English* source repo.

*Reminder: for an editor the source and target are the same org and repo. Thus the adjectives "source" and "target" are dropped in the description below.*

1. Editor selects a file to translate from the list of files from the uW master branch.
2. Branch is created in repo and file is created in the editor's branch and is populated the content from the uW master branch.
3. Both the left (source) side is populated from the repo master branch.
4. The editor's side content may be changed and saved until the editors's work is done.
5. Editor finishes, saves, and a PR is used to merge the updates into the uW master branch.

## Future Catalog Next Process

### Translator Chronology

1. Translator selects a file to translate from the list of files in the uW catalog. The list shown will always be from the latest English production published content in the uW org.
2. Branch is created in target repo and the file is created in a  branch with the English content from the catalog.
3. Translator may continue to make changes and save them until the work is done.
4. Translator finishes, saves, and a PR is used to merge this new file into the target master branch.
5. Later, the translator wants to "edit" the translated file.
6. Again the file is selected from the list of files in the uW catalog.
7. This time, since target master branch has content for this file, it will be used to populate the translator's target branch.
8. The source side is populated from the uW production English catalog.
9. Changes are made, saved, and a PR is used to merge the updates back to the target master branch.

### Editor Chronology

This case describes an update to the uW *English* source repo.

*Reminder: for an editor the source and target are the same org and repo. Thus the adjectives "source" and "target" for "repo" are dropped in the description below.*

1. Editor selects a file to translate from the list of files in the uW production English catalog.  See Note #1 about added content, such as a new Translation Word article.
2. Branch is created in repo and the file is created in a  branch with the English content from the repo.
3. The source side is populated from the uW production English catalog.
4. The Editor may continue to make changes and save them until the work is done.
5. Editor finishes, saves, and a PR is used to merge the changed file into the master branch.

**Note #1** To add new content, the editor must something like this:
1. Create the content manually.
2. Add it directly to the uW English repo master branch. 
3. Once even a partial file is in the master branch and is released to catalog next, then it will be available for further editing within tc-create.

**Comment by mandolyte** [comment]
Note: in branch feature-cn-829-add-catalog-next I have some code that might be usable for a quicker solution.

@ancientTexts-net

**Comment by richmahn** [comment]
> Issue: how to specify Original Language resources (Greek and Hebrew texts).

Probably easiest to just include them in the list as "el-x-koine_ugnt" and "hbo_uhb" and if a resource name like this has an underscore, you assume it already includes the language so you don't prepend a language code.

**Comment by richmahn** [comment]
> If stage is "latest" (default), then...

@mandolyte "latest" stage means the master branch. Don't you mean "prod" as the default? 

Just now proof-reading this.

**Comment by mandolyte** [comment]
@richmahn thanks - I updated the portions above.

**Comment by ancientTexts-net** [comment]
see issue https://app.zenhub.com/workspaces/tc-create-workspace-5f049cdbfa88090016bf043e/issues/unfoldingword/tc-create-app/861

**Comment by ancientTexts-net** [comment]
https://github.com/unfoldingWord/tc-create-app/tree/feature-cn-829-add-catalog-next
feature-cn-829-add-catalog-next

**Comment by ancientTexts-net** [comment]
NOTES:
https://git.door43.org/cecil.new/journals/src/branch/master/tc-create-app/tc-create-app.829.md

**Comment by mandolyte** [comment]
# Testing Notes for QA

## As Editor Role

1. in QA DCS, go to the resource file in unfoldingWord+English, edit it, and propose a change. Proceed with the PR and update master with your change. Now the master branch will be clearly different from the published content. This only needs to be done once after QA is refreshed from production (on Mondays).
2. Delete your branch (if any)
3. Now run tc-create, selecting unfolding word as org, en_tn or en_twl as resource, English as the language, and the file modified in step 1.

*Expected Results*: both source and target will contain the modified data from master branch.

Note: once editing has started, the target will contain whatever is in the users branch at that time.

## As Translator Role

1. in QA DCS, go to the resource file in unfoldingWord+English, edit it, and propose a change. Proceed with the PR and update master with your change. Now the master branch will be clearly different from the published content. This only needs to be done once after QA is refreshed from production (on Mondays).
2. Let's say, you are using Es-419_gl in the translator role and es-419_tn as the resource type. Then delete your branch in that repo if you have one.
3. Now run tc-create, selecting es-419_gl as the org, then TN as the resource type, and es-419 as the language.
4. Select the file modified  in step 1.

*Expected Results*: both source and target will contain content from the latest production release, NOT the master branch from unfoldingWord.

**Comment by mandolyte** [comment]
# Testing

Below are six different use cases to be tested. Altho, since Editor Case A isn't possible, there are actually only five.

In addition to the content shown in the source and target sides, there is also the functionality of the file header chips that must work properly. Both source and target are shown here:

![image](https://user-images.githubusercontent.com/1289437/119657555-119e6900-bdfa-11eb-9e81-305230d40950.png)


The functions are:
- view the license for the repo
- view the DCS webpage for the item
- source: compare vs master
- target: compare vs branch


## Editor Role

This is when unfoldingWord English language resources are being edited.

### Case A: New Item

In this case, an entirely new item is being created. This case is not possible with tc-create. New items must be created using other tools and added to the repository using Gitea DCS itself.

### Case B: Begin Update of Existing Item

This is the common case. It is recognizable when the source and target repos are identical. At present, the source is *always* a repo owned by `unfoldingword`. The following occurs:
- A user branch is created (if it doesn't already exist)
- The current item in the master branch is copied to the user branch, which becomes the starting point for the edits.
- Note that if a second party is also editing, saving, merging the same item, then it is possible that the source side (left) may different from time to time while the first is editing. There is currently no mechanism in place that will notify the first user that this has occurred. Teams must rely on communication to coordinate changes.

### Case C: Continue Update of an Existing Item

In this case, the edits may take several sessions/days/weeks. When the user returns to the item to continue editing:
- No user branch is created, since it already exists
- No content is copied from the master branch to the user branch, since the user edits are being captured there for later merging into the master branch

## Translator Role

This is when other organizations create content, translating from unfoldingWord owned and published resources. The role of translator is recognized by selection of a non-uW organization.

### Case A: New Item

In this case, an entirely new item is being created. The new item, of course, is only new to the translating organization. It will be selected from the `unfoldingword` organization for translation. The source content is currently taken from the latest production uW content. The steps are:
- The source is selected from the uW master branch file list. *There is an edge case where uW might have content in the master branch that is not yet released. This will cause an error when the published version of the content is retrieved.*
- The content for the source is retrieved from the latest published content.
- Note that since a new version of the content may be published while the translator is working, there is at present no mechanism in place to detect and notify the translator that the source from which they are translating may have changed (it may or may not -- it is only a possibility).
- A user branch is created in the translating org's repo and the default content from the uW published content is created there by default. That is, the English content is the starting point for the translation.
- At the end, the source side contains latest uW published content and the target side is identical.
- Eventually, once the translation is done, it will be merged into the org's repo's master branch.

### Case B: Begin Update of Existing Translated Item

In this case, translated content already exists and this will be an update to the existing content.
- A user branch is created in the repo
- The content of the repo's master branch will be used as the starting point will be copied to the user's branch.
- At the end, the source side will have the latest published content from uW and the target side will have the org's repo's master branch content.

### Case C: Continue Update of an Existing Translated Item

In this case, the edits may take several sessions/days/weeks. When the user returns to the item to continue editing:
- No user branch is created, since it already exists
- No content is copied from the master branch to the user branch, since the user edits are being captured there for later merging back into the master branch

**Comment by mandolyte** [comment]
[tc-create-app.829.testing.pdf](https://github.com/unfoldingWord/tc-create-app/files/6548001/tc-create-app.829.testing.pdf)

Testing completed. Not shown in the attachment are the file header link results.

**Comment by BincyJ** [comment]
Related Issues in #822

**Comment by ancientTexts-net** [comment]
# Testing notes
Cecil included some instructions for testing on the Spike:
	https://app.zenhub.com/workspaces/tc-create-workspace-5f049cdbfa88090016bf043e/issues/unfoldingword/tc-create-app/829
	
Cecil's testing notes involve updating MASTER to force a difference between MASTER and the PROD RELEASE VERSION.
This approach is a great idea && probably easier than creating a new release.

But here are some specifics just about where the screens are:

---

# Open Catalog Next Content
- Open English content on DCS
  (unfoldingWord org)
  e.g., https://qa.door43.org/unfoldingWord/en_tn
- Select the **Releases** tab
- Select a Release (e.g., "Version 48)
- Click the "Tag" icon
	to browse release **content**

---

# Create New Release
- Commit some changes to the master branch (in QA! 😲 🙂)
	- Open a content file & click the "Edit" icon
	- Save by clicking "Create a new branch and start a pull request" at the bottom.
		(--or-- begin by editing your own branch)
	- NOTE: you do NOT need to create or merge a PR.
- Create a new release
	- Select the "Releases" tab.
	- Click "New Release"
	- Select the branch you created in the first step.
	- Click "Publish Release"
	- Fill in Version Name and Title fields.
- Now the new release is available to tC Create.

.
